pkg/services/ngalert/schedule/schedule.go

```diff
@@ -162,94 +138,23 @@ func (sch *schedule) Unpause() error {
-	for {
-		select {
-		case ctx := <-evalCh:
-			if evalRunning {
-				continue
-			}
-
-			evaluate := func(attempt int64) error {
-				start := timeNow()
-
-				// fetch latest alert rule version
-				if alertRule == nil || alertRule.Version < ctx.version {
-					q := models.GetAlertRuleByUIDQuery{OrgID: key.OrgID, UID: key.UID}
-					err := sch.ruleStore.GetAlertRuleByUID(&q)
-					if err != nil {
-						sch.log.Error("failed to fetch alert rule", "key", key)
-						return err
-					}
-					alertRule = q.Result
-					sch.log.Debug("new alert rule version fetched", "title", alertRule.Title, "key", key, "version", alertRule.Version)
-				}
-
-				condition := models.Condition{
-					Condition: alertRule.Condition,
-					OrgID:     alertRule.OrgID,
-					Data:      alertRule.Data,
-				}
-				results, err := sch.evaluator.ConditionEval(&condition, ctx.now, sch.dataService)
-				var (
-					end    = timeNow()
-					tenant = fmt.Sprint(alertRule.OrgID)
-					dur    = end.Sub(start).Seconds()
-				)
-
-				sch.metrics.EvalTotal.WithLabelValues(tenant).Inc()
-				sch.metrics.EvalDuration.WithLabelValues(tenant).Observe(dur)
-				if err != nil {
-					sch.metrics.EvalFailures.WithLabelValues(tenant).Inc()
-					// consider saving alert instance on error
-					sch.log.Error("failed to evaluate alert rule", "title", alertRule.Title,
-						"key", key, "attempt", attempt, "now", ctx.now, "duration", end.Sub(start), "error", err)
-					return err
-				}
-
-				processedStates := sch.stateManager.ProcessEvalResults(alertRule, results)
-				sch.saveAlertStates(processedStates)
-				alerts := FromAlertStateToPostableAlerts(sch.log, processedStates, sch.stateManager, sch.appURL)
-				sch.log.Debug("sending alerts to notifier", "count", len(alerts.PostableAlerts), "alerts", alerts.PostableAlerts)
-				err = sch.sendAlerts(alerts)
-				if err != nil {
-					sch.log.Error("failed to put alerts in the notifier", "count", len(alerts.PostableAlerts), "err", err)
-				}
-				return nil
-			}
-
-			func() {
-				evalRunning = true
-				defer func() {
-					evalRunning = false
-					sch.evalApplied(key, ctx.now)
-				}()
+func (sch *schedule) Run(ctx context.Context) error {
+	sch.wg.Add(1)
 
-				for attempt = 0; attempt < sch.maxAttempts; attempt++ {
-					err := evaluate(attempt)
-					if err == nil {
-						break
-					}
-				}
-			}()
-		case <-stopCh:
-			sch.stopApplied(key)
-			sch.log.Debug("stopping alert rule routine", "key", key)
-			// interrupt evaluation if it's running
-			return nil
-		case <-grafanaCtx.Done():
-			return grafanaCtx.Err()
+	go func() {
+		if err := sch.ruleEvaluationLoop(ctx); err != nil {
+			sch.log.Error("failure while running the rule evaluation loop", "err", err)
 		}
-	}
+	}()
+
+	sch.wg.Wait()
+	return nil
 }
 
-func (sch *schedule) Ticker(grafanaCtx context.Context) error {
-	dispatcherGroup, ctx := errgroup.WithContext(grafanaCtx)
+func (sch *schedule) ruleEvaluationLoop(ctx context.Context) error {
```

Comment:
This is the "big" part of the refactor:  Moved the `Ticker` code into this function so that I can discretely call it into a new go-routine

Reply:
My understanding here is that on each tick, we'll run X amount of rules, spawning a new routine per rule means we don't block and evaluate them all at the same time to avoid delays.

In a nutshell, the benefit seems to be faster evaluations?

Reply:
sorry, I had totally overseen this:
> We'll introduce another routine that checks the database for
configuration and spawns other routines accordingly.